Update AWS CloudTrail module to use boto3

shepdelacreme opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Rewrite of the AWS CloudTrail module to support the following options/parameters that are not currently supported in the module:

- Tagging of the resource
- Enabling/disabling Multi-region trail
- Starting and stopping logging on the trail at will
- Enabling/disabling log file validation feature for delivered log files
- Enabling/disabling log file encryption using KMS key for delivered log files 
- Setting/removing an SNS Topic for log delivery notifications
- Enabling/disabling event delivery to a CloudWatch Logs log group

The boto2 modules for CloudTrail do not support the above features so this module was rewritten to use boto3.

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
lib/ansible/modules/cloud/amazon/cloudtrail.py

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.4.0 (devel 9ad62a48ab) last updated 2017/03/22 12:04:36 (GMT -400)
  config file = /Users/username/.ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Dec 18 2016, 07:03:39) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

Commit:
rewrite AWS CloudTrail module to use boto3 in order to support updated/new CT parameters not supported with boto2

remove cloudtrail.py from legacy-files as it passes pep8 tests now

update return values to use snake case per coding guidelines

update boto3 exception handling to use method outlined in coding guidelines

update parameter spec and return value code per PR code review

mattclay commented:
Closing and re-opening to trigger CI.

shepdelacreme commented:
I'm not really sure how this process works so I'll ask. Should I rebase my changes against devel every time there is an update to devel or wait until the code is reviewed, then rebase? i.e. Is it worth the effort to have it merge ready at all times or wait till a reviewer says "yes we want this change - now go rebase"?

mattclay commented:
@shepdelacreme Generally you only need to rebase when there are merge conflicts with your PR, as there currently are. Most of the time updates to devel won't cause merge conflicts, but they do happen, particularly on modules which see more frequent updates.

However, if there is a question as to whether or not your PR will be accepted at all, you may want to hold off on rebasing until you've discussed your changes with the appropriate maintainers.

shepdelacreme commented:
@mattclay Thanks. I just squashed my commits and rebased.

I know this is a very busy project so apologies for asking but having not done this before I don't know. Do I need to reach out to specific maintainers or a specific mailing list to say "hey here is a PR for a community module that needs review" or is the preferred method to just submit the PR and wait in line. My aim in asking is to not come across as pushy and make sure I'm respecting everyone that works on this projects time!

Thanks again!

mattclay commented:
@shepdelacreme Take a look at the [bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md) for information on our PR workflow.

If you would like to discuss your PR, the best place is usually our #ansible-devel IRC channel on Freenode. Since this is a cloud module, you probably want to talk to @s-hertel (shertel on IRC).

s-hertel commented:
Hi, @shepdelacreme. Let me know if I can do anything to help you out! :) I'm excited things are being updated to use Boto3.

shepdelacreme commented:
@s-hertel Thanks! Let me know if you have any questions during review. It is nearly a complete rewrite but I decided to keep the interface compatible with the old module that used `state=enabled` and `state=disabled` instead of the more standard `present/absent` so any existing tasks shouldn't break upon module update.

willthames commented:
@shepdelacreme here to help too.

I would advise just allowing state to be present, absent, enabled or disabled. Less surprising

willthames commented:
bot_status

willthames commented on lib/ansible/modules/cloud/amazon/cloudtrail.py:
```diff
@@ -19,231 +19,583 @@
+    type: complex
+    sample: hash/dictionary of values
+    contains:
+        TrailARN:
```
This should be trail_arn.

The camel_dict_to_snake_dict should be used to convert boto3 objects to ansible objects before returning them through module.exit_json

willthames commented on lib/ansible/modules/cloud/amazon/cloudtrail.py:
```diff
@@ -19,231 +19,583 @@
+        if module.check_mode:
+            trail = dict()
+            trail.update(ct_params)
+            trail.update({'LogFileValidationEnabled': ct_params['EnableLogFileValidation']})
```
use `trail['LogFileValidationEnabled'] = ct_params['EnableLogFileValidation']`

willthames commented on lib/ansible/modules/cloud/amazon/cloudtrail.py:
```diff
@@ -19,231 +19,583 @@
+        name=dict(default='default', type='str'),
+        enable_logging=dict(default=True, type='bool', required=False),
+        s3_bucket_name=dict(type='str', required=False),
+        s3_key_prefix=dict(type='str', required=False),
```
`required=False` is the default, and no not required.

Same for `type=str`

willthames commented on lib/ansible/modules/cloud/amazon/cloudtrail.py:
```diff
@@ -19,231 +19,583 @@
+    try:
+        client.update_trail(**ct_params)
+    except ClientError as err:
+        module.fail_json(msg=str(err))
```
Use standard boto3 exception handling: 

- https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/GUIDELINES.md#boto3-2

I've just noticed there's a missing `c` at the end of `traceback.format_exc()`, so just be careful  there :) - I've added #23168 to fix that

shepdelacreme commented:
@willthames Made the requested updates and rebased. Thanks for the review!

ready_for_review

ryansb commented:
cc/ @tedder

tedder commented:
`shipit`. Thanks for rewriting.

s-hertel commented:
shipit

shepdelacreme commented:
Is there a reason this hasn't automerged yet even though it has the correct number of approvals? Do I need to rebase or something?

s-hertel commented:
@shepdelacreme Because it touches more than just a module file.

s-hertel commented:
@willthames can you clear your review?

s-hertel commented:
bot_status
changes_requested
